### ISSUE TITLE
Perspective correction tool

### ISSUE THREAD

**Issue by lightsilverberryfox** [vigri/scantailor-advanced#28]
sometimes when you want to scan something, the pictures are not square, and that causes lots of issues with detect content.
i suggest that we have a tool similar to GIMP's perspective tool to correct for it.

is this still maintained @vigri ?

**Comment by darnn** [comment]
I suspect this may be beyond the scope of ScanTailor in general, especially since, as you've pointed out, there are other programs that already do this. Also, this is not dissimilar to what dewarping does at the output stage.

**Comment by lightsilverberryfox** [comment]
@darnn I don't think that it's beyond the scope of ScanTailor because, as I already said, non-square images and detect content get along badly, mainly because the detect content tool does not allow free-from selection, only square.

While it is true that you may use other tools like GIMP but perspective issues usually apply to every page and I don't think GIMP perspective tool has batch processing, besides, as you already said, it's basically moving the dewarping to the preprocessing phase.

**Comment by darnn** [comment]
Maybe I'm not quite understanding you, then. What do you mean when you say that you scan something and pictures aren't square? Is there a specific example of a scan you can link (or upload)?

**Comment by lightsilverberryfox** [comment]
@darnn No, I do not have any examples that I can upload albeit, it's very easy to see the effect by taking a document, capturing a picture at an odd angle and give the picture to scan tailor.

**Comment by darnn** [comment]
Okay, I think that makes it a bit clearer at least: You're talking about using a camera, rather than a scanner? Because I do believe that in most situations where people use cameras, they have some kind of line-straightening process (or what have you) that they apply before they start things like binarization. But I admit I have limited experience with that.
Anyway, this is all mostly academic, since as far as I understand, this repository mostly exists to have a functional build of the latest version of STA, and there aren't many additions planned, much as I would like some myself. But hey, shoot your shot.

**Comment by lightsilverberryfox** [comment]
Yes, I'm talking about using a camera, using a scanner with ScanTailor doesn't seem to have much sense to me because the scanner already does processing and Scan Tailor is not needed.

**Comment by darnn** [comment]
None of the scanners I've had have done processing nearly as well as ScanTailor... right now I have a copier, which is great for speed but not in any other regard, so I at least definitely do need it for scans, but I see what you mean. Are you using CHDK or anything like that, or just taking photos by hand?

**Comment by lightsilverberryfox** [comment]
I take pictures by hand, with a tripod and with manual trigger. It's not perfect but my camera doesn't have support for CHDK. If the perspective is not perfect, it's very hard or sometimes impossible to crop and select the content.

**Comment by darnn** [comment]
Ah. Well, the only thing I can suggest is, you can try to replicate the flattening of the page by the scanner by pressing the book down using a large glass pane. I've used my phone to "scan" a few books this way, and it was much easier with the glass pane than trying to hold them open (and straight) by hand. Though of course then you get issues like glare, but I find that it's mostly okay if you figure out the proper angles.
For instance, here's a not-quite-straight image from when I was holding the book open myself (taken by vFlat, which does some processing and straightening of its own):
![20210202_203600000_ - Copy](https://user-images.githubusercontent.com/19672444/202897643-bd2307db-f218-4867-b96f-ede40fe841a4.jpg)

And here's one (also taken using vFlat) with a glass pane:
![20210310_150540000_ - Copy](https://user-images.githubusercontent.com/19672444/202897715-ee21fb6e-bf45-4b8e-b980-c1a444b1a955.jpg)

As you can see, still not totally straight, but much better. And there's no glare since I worked out the angle ahead of time.

In any case, if you're planning on digitizing many books over time, I really recommend looking into a cheap second-hand copier with scanning capabilities. I cannot overstate what a huge difference it makes in terms of effort and speed.

**Comment by lightsilverberryfox** [comment]
@darnn This is the scanner i use: https://www.instructables.com/DIY-High-Speed-Book-Scanner-from-Trash-and-Cheap-C/

As you can see it's just made from cardboard, but getting the camera aligned properly is very hard, unfortunately, better DIY scanners cost around 200$ which is expensive. This is why a perspective correction tool would be very very helpful for this cases.

**Comment by darnn** [comment]
Ah, well, if that's your price range, then yes, I think this might be a good compromise. Anyway, if I had to do this to a large number of images, I would find some way of obtaining Photoshop, since it does have a batch mode. And GIMP does have a batch plugin of some sort (BIMP or something), though I admit I haven't tried it.
The quick and dirty thing I use for this is YASW: https://sourceforge.net/projects/yascanw/
But as you can see, it's both old and fairly rudimentary. It does seem like it was created for your exact purpose, though, so that's something.

**Comment by pablogventura** [comment]
**Perspective correction** (GIMP-style) is a **large feature** — still open; related requests exist on other forks.

No ETA here. If you start a prototype, link the branch early for feedback.

Keeping open as the main tracking issue for this repo.
